storage/replica_command.go

```diff
@@ -646,6 +650,13 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
 func (r *Replica) HeartbeatTxn(batch engine.Engine, ms *engine.MVCCStats, args proto.HeartbeatTxnRequest) (proto.HeartbeatTxnResponse, error) {
 	var reply proto.HeartbeatTxnResponse
 
+	if args.Txn == nil {
```

Comment:
Perhaps we should write some tests that basically recreate all kinds of bad sender scenarios, and then add code that protects us. That could be a slightly more ambitious change. Thanks for taking this on!